jupyterhub/utils.py

```diff
@@ -82,13 +94,51 @@ def can_connect(ip, port):
         return True
 
 
-def make_ssl_context(keyfile, certfile, cafile=None, verify=True, check_hostname=True):
-    """Setup context for starting an https server or making requests over ssl."""
+def make_ssl_context(
+    keyfile,
+    certfile,
+    cafile=None,
+    verify=None,
+    check_hostname=None,
+    purpose=ssl.Purpose.SERVER_AUTH,
+):
+    """Setup context for starting an https server or making requests over ssl.
+
+    Used for verifying internal ssl connections.
+    Certificates are always verified in both directions.
+    Hostnames are checked for client sockets.
+
+    Client sockets are created with `purpose=ssl.Purpose.SERVER_AUTH` (default),
+    Server sockets are created with `purpose=ssl.Purpose.CLIENT_AUTH`.
+    """
     if not keyfile or not certfile:
         return None
-    purpose = ssl.Purpose.SERVER_AUTH if verify else ssl.Purpose.CLIENT_AUTH
+    if verify is not None:
+        purpose = ssl.Purpose.SERVER_AUTH if verify else ssl.Purpose.CLIENT_AUTH
+        warnings.warn(
+            f"make_ssl_context(verify={verify}) is deprecated in jupyterhub 2.4."
+            f" Use make_ssl_context(purpose={purpose}).",
+            DeprecationWarning,
+            stacklevel=2,
+        )
+    if check_hostname is not None:
+        purpose = ssl.Purpose.SERVER_AUTH if check_hostname else ssl.Purpose.CLIENT_AUTH
+        warnings.warn(
+            f"make_ssl_context(check_hostname={check_hostname}) is deprecated in jupyterhub 2.4."
+            f" Use make_ssl_context(purpose={purpose}).",
```

Comment:
I wondered how those objects would render like this, maybe rendering them as `ssl.Purpose.SERVER_AUTH` for example would make it clear. This change suggestion makes us render it like that.

![image](https://user-images.githubusercontent.com/3837114/181344820-fda03e6e-f001-474d-9c9a-48af727b5d13.png)

```suggestion
            f" Use make_ssl_context(purpose=ssl.{str(purpose)}).",
            DeprecationWarning,
            stacklevel=2,
        )
    if check_hostname is not None:
        purpose = ssl.Purpose.SERVER_AUTH if check_hostname else ssl.Purpose.CLIENT_AUTH
        warnings.warn(
            f"make_ssl_context(check_hostname={check_hostname}) is deprecated in jupyterhub 2.4."
            f" Use make_ssl_context(purpose=ssl.{str(purpose)}).",
```

Reply:
Went with `{purpose!s}`, which is the standard way to specify str-format conversion. I thought this was the default, but apparently it's `__format__()` which is [complex for enums](https://github.com/python/cpython/blob/v3.10.0/Lib/enum.py#L750-L767) (_most_ enums would produce `ssl.Purpose.SERVER_AUTH`, but not ssl.Purpose for weird reasons to do with `_member_type_`). After digging into why this wasn't what I expected, I also learned that the simpler behavior [has become the default in 3.11](https://github.com/python/cpython/blob/v3.11.0b5/Lib/enum.py#L1204-L1205).

I also just learned _today_ that it's `__format__` not `__str__` that sets the default substitution in f-strings and other similar situations. This has been true since Python 2.6!